source/extensions/filters/network/thrift_proxy/protocol.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <memory>
+#include <stack>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+#include "common/singleton/const_singleton.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+/**
+ * Names of available Protocol implementations.
+ */
+class ProtocolNameValues {
+public:
+  // Binary protocol
+  const std::string BINARY = "binary";
+
+  // Lax Binary protocol
+  const std::string LAX_BINARY = "binary/non-strict";
+
+  // Compact protocol
+  const std::string COMPACT = "compact";
+
+  // JSON protocol
+  const std::string JSON = "json";
+
+  // Auto-detection protocol
+  const std::string AUTO = "auto";
+};
+
+typedef ConstSingleton<ProtocolNameValues> ProtocolNames;
+
+/**
+ * Thrift protocol message types.
+ * See https://github.com/apache/thrift/blob/master/lib/cpp/src/thrift/protocol/TProtocol.h
+ */
+enum MessageType {
```

Comment:
I recommend using `enum class` rather than plain `enum` here, so that the value names (Call, Reply, etc) get namespaced within `MessageType`.